Account for additional `cluster_type` in clustering calculations script

cbethell opened the pull request:
**Issue Addressed**
Closes #212 

**What is the purpose of these changes?**
<!--Provide some background on the changes proposed.-->
This PR adds an additional `cluster_type` parameter to the `optional-clustering-analysis/clustering-calcutions.R` script, as to allow for users to compare more than one `cluster_type` across the same range of parameters.

**What changes did you make?**
<!--Describe the concrete changes that you made or additional features that were added, be as detailed as possible in the steps you took.-->
More specifically, this PR:

1. Adds an optparse param to the `clustering-calculations.R` script named `second_cluster_type` 
2. Checks if there is a second cluster type provided and performs clustering for that type if it provided
3. Creates wrapper functions to check the cluster validity and stability stats, as to have all of the cluster stats (for all specified cluster types) in each relevant output file

This PR also removes the 2.18 specification associated with the pandoc required dependency, due to the following error and also related to https://github.com/AlexsLemonade/scpca-downstream-analyses/pull/222#pullrequestreview-1118636195:

<img width="1125" alt="Screen Shot 2022-09-23 at 9 58 52 AM" src="https://user-images.githubusercontent.com/43576623/192886787-5e6725dd-3e9b-42c1-bedb-98781e639edd.png">

Removing the "2.18" allowed things to run successfully.

**Edit:** mostly fixed the above error so reverted the pandoc change here

**Any comments, concerns, or questions important for reviewers**
- Does the code added in this PR seem like the most suitable way to tackle #212? Do you have suggestions for refactoring?
- 
**Checklist**

Place an `x` in all boxes that you have completed.

- [x] I have run the most recent version of the code
- [ ] If I am adding in reports or generating any plots, I have reviewed the necessary reports
- [ ] I have added all necessary documentation (if applicable)

allyhawkins commented on optional-clustering-analysis/clustering-calculations.R:
```diff
@@ -169,47 +176,128 @@ nn_range <- define_nn_range(opt$nearest_neighbors_min,
-validity_stats_df <- create_metadata_stats_df(sce, 
-                                              nn_range, 
-                                              opt$cluster_type)
+cluster_stability_stats_wrapper_function <- function(sce, nn_range, cluster_type, second_cluster_type = NULL, ...) {
+  # Purpose: Calculate the cluster stability stats for each provided `cluster_type`
+  # across n range of parameters
+  #
+  # Args:
+  #   sce: clustered SingleCellExperiment object
+  #   nn_range: the range of numeric parameters used to test for clustering
+  #   cluster_type: method used for clustering
+  #   second_cluster_type: additional method used for clustering
+
+  # Check the cluster stability stats and return the summary ari in a data frame
+  summary_stability_stats_df <-
+    get_cluster_stability_summary(sce,
+                                  nn_range,
+                                  cluster_type) %>%
+    dplyr::mutate(param_value = as.numeric(param_value))
+  
+  # If a second cluster type is provided, check the cluster stability stats for
+  # each of the clusters in the SCE object associated with the second cluster type
+  if (!is.null(second_cluster_type)) {
+    second_summary_stability_stats_df <- 
+      get_cluster_stability_summary(sce,
+                                    nn_range,
+                                    second_cluster_type) %>%
+      dplyr::mutate(param_value = as.numeric(param_value))
+    
+    summary_stability_stats_df <- summary_stability_stats_df %>%
+      dplyr::bind_rows(second_summary_stability_stats_df)
+  }
+}
```
```suggestion
  # Check the cluster stability stats and return the summary ari in a data frame
  summary_stability_stats_df_list <- purrr::map(cluster_types, 
    ~ get_cluster_stability_summary(sce,
                                  nn_range,
                                  .x )
 summary_stability_stats_df <- dplyr::bind_rows(summary_stability_stats_df) &>&
    dplyr::mutate(param_value = as.numeric(param_value))
 
```
Rather than have a whole other function you can condense this and use `purrr` here to make a list of dataframes, one for each cluster type and then combine together into one dataframe.

allyhawkins commented on optional-clustering-analysis/clustering-calculations.R:
```diff
@@ -169,47 +176,128 @@ nn_range <- define_nn_range(opt$nearest_neighbors_min,
+  }
+}
+
+# Run the cluster validity stats wrapper function
+validity_stats_df <- cluster_validity_stats_wrapper_function(
+  sce, nn_range, opt$cluster_type, 
+  second_cluster_type = opt$second_cluster_type)
```
```suggestion
```

allyhawkins reviewed (commented):
Thanks for getting this started. My overall comment is to actually use a single argument instead of two and read in a string of cluster options and separate that into a vector. You can then loop through that vector of cluster options and apply the functions that you have written. This should hopefully simplify things and also make things more robust to any future changes like additional clustering we would want to do.

cbethell commented:
Thanks for the helpful feedback @allyhawkins! I was able to revert back to one `cluster_type` param option, supply the vector of multiple cluster types there, and use `str_split()` to handle those cluster types individually. I also added the line to the Snakefile in the last commit https://github.com/AlexsLemonade/scpca-downstream-analyses/pull/224/commits/c6b2f0ae7f71f5e121ee8c6246e82f9705e52ae6 that gets the modifications in this PR working successfully. Noting that I added a `cluster_types` parameter to the config.yaml file since we use `cluster_type` in the core workflow and only expected one cluster type there -- let me know what you think here and overall!

allyhawkins reviewed (approved):
LGTM!

sjspielman commented on config.yaml:
```diff
@@ -31,6 +31,7 @@ n_genes_pca: 2000 # The `n` number of highly variable genes to subset for PCA
 
 # clustering parameters
 cluster_type: "louvain"
```
This line should be removed now, right?

sjspielman replied:
Ah, I see. The names are a bit to close for comfort. Seems really easy to use the wrong one. How about `cluster_types` --> `opt_cluster_types` (like "optional")? I'm not sure about that either though. Tagging @allyhawkins back in for ideas!

allyhawkins replied:
Hmm looking at this I have a few ideas. Both of which would include changing the name. I would say maybe we need to change both argument names so that we have `core_cluster_type` and `optional_cluster_types` or something like that? So that we are more explicit? 

Additionally I think we can do some reorganization but that could go one of two ways.  My first ideas is to actually split out any of the clustering specific parameters into their own config file that are parameters that just get accessed in this clustering exploration module. Then you can call both config files at the beginning of the snakefile with something like this (assuming that all the config files also get moved to a new config directory which I would also suggest).
```
configfile = "config/config.yaml"
configfile = "config/cluster_config.yaml"
```

The other option is to just move all the optional clustering parameters to another section in the larger main configfile with a comment above explaining that they are for that module. We want to make sure that users understand where each parameter goes. 

I think I favor the first option because someone using just the core workflow might be confused as to why there are these extra parameters that are irrelevant to what they are using. Additionally we are going to have more modules in the future so we will probably want to have config files that have module specific parameters moving forward.

sjspielman replied:
I like `core_cluster_type` and `optional_cluster_types` as explicit parameter names.

Just gonna also note that I find that `Snakemake` behavior WEIRD. I would expect writing `configfile` twice to overwrite it, but apparently not ¯\_(ツ)_/¯

allyhawkins replied:
>Just gonna also note that I find that Snakemake behavior WEIRD. I would expect writing configfile twice to overwrite it, but apparently not 🤷‍♀️

It would overwrite it if you had conflicting parameter names I think, but if they are unique parameters then it should use all of them as far as I can tell in the docs.

sjspielman replied:
that makes some sense I suppose..

jashapiro replied:
I'll weigh in with a slight difference: I would leave `config.yaml` in root directory, and only move `cluster_config.yaml` to the subdirectory, which I might name `opt_configs` or something like that.

And a note on syntax, which might make @sjspielman a bit happier:
```
configfile: "config.yaml"
configfile: "opt_configs/cluster_config.yaml"
```

(Note the colon instead of `=`)

cbethell replied:
As Josh and I discussed in science pair it, I have made personal note of his comment above in case it comes up in discussion later, but will leave things as is for now!

sjspielman commented on optional-clustering-analysis/clustering-calculations.R:
```diff
@@ -138,9 +138,14 @@ if (!file.exists(opt$sce)){
-if (!opt$cluster_type %in% c("louvain", "walktrap")) {
-  stop("--cluster_type (-c) must be either louvain or walktrap.")
+# Split up string of cluster types
+cluster_types <- unlist(stringr::str_split(opt$cluster_types, ","))
```
If you want to protect against someone adding a space, e.g.`louvain, walktrap`, we can do:


```suggestion
cluster_types <-  stringr::str_split(opt$cluster_types, ",") %>% unlist() %>% stringr::str_trim()
```

sjspielman commented on optional-clustering-analysis/clustering-calculations.R:
```diff
@@ -138,9 +138,14 @@ if (!file.exists(opt$sce)){
+# Check that clustering type is valid
+for (cluster_type in cluster_types) {
+  if (!cluster_type %in% c("louvain", "walktrap")) {
+    stop("--cluster_types (-c) must be either louvain or walktrap.")
```
Update this message - it can be either or BOTH, comma separated.

sjspielman commented on optional-clustering-analysis/clustering-calculations.R:
```diff
@@ -169,47 +174,64 @@ nn_range <- define_nn_range(opt$nearest_neighbors_min,
-existing_columns <- intersect(cluster_column_names, colnames(colData(sce)))
-
-if(length(existing_columns) != 0){
+cluster_wrapper_function <- function(sce, nn_range, cluster_type, ...) {
```
I generally like for function names to be more action-oriented, which is a preference I get from the style guide: https://style.tidyverse.org/functions.html#naming 

In this case, mostly I don't like that `function` is in the name of the function. We know it's a function, but what does it do? It performs the clustering, and incidentally does so by wrapping other functions. 
How do you feel about naming it (and updating its use below) `perform_clustering()`?

cbethell commented:
Thank you both for the feedback @sjspielman and @allyhawkins! I have created separated out the parameters to have two config files as Ally suggested, renamed the parameters to be more explicit, and successfully tested out the changes! Let me know what you think now!

cbethell commented:
> Changes look good to me! The only other suggestion I have is to update the README with this new parameter setup. There are several places where this should be changed throughout.

Sweet, thank you! I am planning to put the README updates in a subsequent PR -- likely the same PR that will be tackling https://github.com/AlexsLemonade/scpca-downstream-analyses/issues/218!

sjspielman reviewed (approved):
👍
